bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/BulkReadVeneerApi.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.hbase.wrappers.veneer;
+
+import com.google.api.core.ApiFunction;
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.InternalApi;
+import com.google.api.gax.batching.Batcher;
+import com.google.bigtable.v2.ReadRowsRequest;
+import com.google.bigtable.v2.RowFilter;
+import com.google.cloud.bigtable.data.v2.BigtableDataClient;
+import com.google.cloud.bigtable.data.v2.internal.RequestContext;
+import com.google.cloud.bigtable.data.v2.models.Filters;
+import com.google.cloud.bigtable.data.v2.models.Query;
+import com.google.cloud.bigtable.data.v2.models.Row;
+import com.google.cloud.bigtable.hbase.adapters.Adapters;
+import com.google.cloud.bigtable.hbase.wrappers.BulkReadWrapper;
+import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.protobuf.ByteString;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.apache.hadoop.hbase.client.Result;
+
+/** For internal use only - public for technical reasons. */
+@InternalApi("For internal usage only")
+public class BulkReadVeneerApi implements BulkReadWrapper {
+
+  private static final BlockingQueue<RowFilter> BATCHER_QUEUE = new LinkedBlockingQueue<>();
+
+  private final ScheduledExecutorService batcherPool;
+  private final String tableId;
+  private final BigtableDataClient dataClient;
+  private final RequestContext requestContext;
+  private final Map<RowFilter, Batcher<ByteString, Row>> filterToBatcher;
+
+  // Needs this to stop race condition b/w batch.close()
+  private final Object lock = new Object();
+
+  BulkReadVeneerApi(String tableId, BigtableDataClient dataClient) {
+    this.tableId = tableId;
+    this.dataClient = dataClient;
+    this.filterToBatcher = new HashMap<>();
+
+    // required this to retrieve rowKey and filter from Query instance
+    this.requestContext = RequestContext.create("", "", "");
+
+    // Created pool as instance variable as users are not necessarily use bulkRead every time.
+    this.batcherPool = Executors.newSingleThreadScheduledExecutor();
```

Comment:
When does this get shutdown? I think you are leaking threads here. I think having a single thread is preferable